feat: replace standard logging with structured logging

therealvio opened the pull request:
# Purpose 🎯 

These changes replace the standard library logging with structured logging using [structlog](https://www.structlog.org/en/stable/index.html). Structured logging provides easier-to-parse capabilities for toolings and humans when going through logs.




# Context 🧠 

- Part 1 to satisfy https://github.com/chime/terraform-aws-alternat/issues/76
  - I intend to contribute a follow-up PR that adds logging all connection test events for the purpose of fostering greater support for SLIs, as mentioned in [this comment](https://github.com/chime/terraform-aws-alternat/issues/76#issuecomment-1644817947). For those unfamiliar, calculating the SLI is a matter of `bad events/all valid events`, and right now `aws-alternat` exercises a kind of "no news is good news" reporting.

# Notes 📓 

- I have performed an in-place replacement of the default logger
- No changes to the message contents have been, with some exceptions that have been explicitly pulled out into standalone commits for ease of reviewing.
- Example screenshot of debug log messages through `structlog`
![image](https://github.com/user-attachments/assets/13d58280-dcfd-4685-841e-29463cab6edc)

therealvio committed:
feat: use structured logging with structlog

therealvio commented on functions/replace-route/app.py:
```diff
@@ -4,17 +4,34 @@
+
+# logger is still needed to set the level for dependencies
 logger = logging.getLogger()
-logger.setLevel(logging.INFO)
 logging.getLogger('boto3').setLevel(logging.CRITICAL)
 logging.getLogger('botocore').setLevel(logging.CRITICAL)
```
I found [this guide](https://www.structlog.org/en/stable/standard-library.html) with regards to working with the standard library, but I wasn't sure how I could make this applicable here. Otherwise, third party deps will continue using the standard logger unless someone can figure out if wrapping the `logger` is possible.

bwhaley replied:
My colleague suggests that we could use https://nhairs.github.io/python-json-logger/latest/ as the formatter for these handlers. e.g. something like:

```
handler = logging.StreamHandler()
handler.setFormatter(JsonFormatter())
logging.getLogger('boto3').addHandler(handler)
logging.getLogger('botocore').addHandler(handler)
```

Do you think this would work?

therealvio replied:
I think so! This would align with that guide, the section specifically being: "[Rendering using logging-based formatters](https://www.structlog.org/en/stable/standard-library.html#rendering-using-logging-based-formatters)".

I will give it a go and push out a separate commit for this.

therealvio replied:
An FYI: In the change, `handler` is a loaded term, since that's the entry point of the function. For now, I have opted for `lhandler` and we can come up with a better name during as we iterate and review.

therealvio commented on functions/replace-route/app.py:
```diff
@@ -4,17 +4,34 @@
 
-
+slogger = structlog.get_logger()
+# use structlog's production-ready, performant example config
+# ref: https://www.structlog.org/en/stable/performance.html#example
```
Talking to this particular config, when I ran into the example in [this document](https://www.structlog.org/en/stable/performance.html), it matched what I had in my head as a starting point. The company I work at have some opinions on things, though I wasn't going to impose them here.

therealvio commented on functions/replace-route/app.py:
```diff
@@ -4,17 +4,34 @@
+        structlog.processors.add_log_level,
+        structlog.processors.format_exc_info,
+        structlog.processors.TimeStamper(fmt="iso", utc=True),
+        structlog.processors.EventRenamer("message"),
```
By default, this value is "event" though that term seems to be loaded since the Lambda accepts "events" and I thought that maybe `message` would be a good compromise. Though I am wary of long-term implications of this choice.

Either way, I don't have a strong opinion on whether it stays as this, we leave it as the default `event`, or we use a different name.

therealvio replied:
Then `message` it shall be 🙌

therealvio commented on functions/replace-route/app.py:
```diff
@@ -163,20 +180,20 @@ def check_connection(check_urls):
     if not isinstance(event, dict):
-        logger.error(f"Unknown event: {event}")
+        slogger.error("Unknown event", eventPayload=event)
         return
 
     if event.get("source") != "aws.events":
-        logger.error(f"Unable to handle unknown event type: {json.dumps(event)}")
+        slogger.error("Unable to handle unknown event type", eventPayload=json.dumps(event))
```
As the diff indicates, this changed from a string interpolation to a key. As [this comment](https://github.com/chime/terraform-aws-alternat/pull/122/files#r1916088562) alludes to, the term `event` is loaded.

I am inclined to drop this key and keep the string interpolation like I did everywhere else, or do a combination of both, or even leave that decision to another PR.

therealvio replied:
I decided to preserve the string interpolation in a later commit and will leave changes to the log message contents themselves to a later PR.

therealvio committed:
fix: use slogger as drop-in replacement

This changes removes the addition of the eventPayload key that was
imposed in the previous commit. I would rather leave the decision to the
maintainers.

Whether it be reverting *this* commit, or accepting this one, or leaving
it to the future PR.

bwhaley commented:
Thank you for the PR! I am a bit busy atm, but I hope to review & discuss in the coming days. Appreciate the patience!

bwhaley committed:
Trigger tests

bwhaley commented on functions/replace-route/app.py:
```diff
@@ -4,17 +4,34 @@
+)
+
+# logger is still needed to set the level for dependencies
 logger = logging.getLogger()
```
Is this line still needed?

therealvio replied:
The snippet is selecting 4 lines, I gather this may be for the comment.

This relates to [this comment](https://github.com/chime/terraform-aws-alternat/pull/122#discussion_r1916077063) regarding third party dependencies. `structlog` doesn't appear to completely drop-in to replace `logger` in this specific area. If it's not valuable, I am happy to drop it. My intention here was to describe why `logger` is still being used while `slogger` is in place.

I don't believe any of my commits, or this PR actually describes why they're staying, so a `git log` wouldn't help anyone in that area :(. I don't know if the commit will be a squash or not, but if it's a merge commit, then I can amend the commit message where the surrounding change happened if it helps?

bwhaley replied:
What I'm confused about is that logger = logging.getLogger() is still set here but is never used throughout the file. The levels are set using logging.getLogger below. But is the logger object still used in some way? I may be misunderstanding.

therealvio replied:
Looking back on this, and recollecting:

What's going on here is described in the [structlog docs](https://www.structlog.org/en/stable/standard-library.html) regarding configuring the stdlib `logger` being used by third party dependencies like `boto` and `botocore` that use stdlib `logger`. In this case, we are configuring the log level for these dependencies.

I believe we solved the JSON formatting problem for a similar problem in [this commentary thread](https://github.com/chime/terraform-aws-alternat/pull/122#discussion_r1929392097). This is similar in nature.

bwhaley replied:
Yep, that makes sense.

bwhaley reviewed (commented):
Left a couple comments/ideas to consider.

Mind taking a peek at `test_replace_route.py` and adding a unit test?

bwhaley commented on functions/replace-route/app.py:
```diff
@@ -120,10 +137,10 @@ def replace_route(route_table_id, nat_gateway_id):
     }
     try:
-        logger.info("Replacing existing route %s for route table %s", route_table_id, new_route_table)
+        slogger.info("Replacing existing route %s for route table %s", route_table_id, new_route_table)
```
Also, most folks should be monitoring for `ERROR` condition or the "Failed connectivity tests! Replacing route" message. This one is `INFO`.

therealvio replied:
No worries.

> Also, most folks should be monitoring for ERROR status or the "Failed connectivity tests! Replacing route" message. This one is INFO.

You'd think it would go without saying. Perhaps in the release we could include that as a sub dot-point? It feels weird having to tell people "how" to monitor things, but mistakes happen.

therealvio replied:
I have made this into a standalone commit including a breaking change footer for the commit. 

I had to make some tweaks to your suggested log line because `NAT Gateway ID` assumes only NAT gateway IDs are being fed in. When the `replace_route()` function is called, it can use either a NAT gateway or EC2 instance ID. So instead, I reworded it as: `Updating route table %s to use NAT target %s"`.

therealvio commented:
Thanks for the feedback thus far, I will go through these as I get time, but it likely won't be until next week if that's cool?

therealvio commented on functions/replace-route/app.py:
```diff
@@ -163,20 +180,20 @@ def check_connection(check_urls):
 
     for rtb in route_tables:
         replace_route(rtb, nat_gateway_id)
-        logger.info("Route replacement succeeded")
+        slogger.info("Route replacement succeeded")
     return False
 
 
 def connectivity_test_handler(event, context):
     if not isinstance(event, dict):
-        logger.error(f"Unknown event: {event}")
+        slogger.error("Unknown event: %s", {event})
```
No worries!

therealvio commented:
> Mind taking a peek at test_replace_route.py and adding a unit test?

I may be missing something, was there something in particular you were looking for me to add?

Full transparency here: I never got the test suite to run, mostly because I never did testing in the Python ecosystem before. I am more than keen to check out any material that would be help here. As long as I can get it running, I can extend the test suite without a fuss :)

For what it is worth: I mostly tested through running the SAM app locally, which served me well enough for this PR.

bwhaley commented:
The two things to add would be tests and the potential use of python-json-logger.

You can look at the GitHub workflow to see how the tests run. It's basically this
```
          pip install pip --upgrade
          pip install pyopenssl --upgrade
          pip install -r functions/replace-route/requirements.txt
          pip install -r functions/replace-route/tests/test_requirements.txt
          python -m pytest
```

therealvio commented:
To keep you in the loop: I am going to be slammed next week, and I am not sure if I will get time to take a look this week.

This is still on my mind, it just may be a while before I get back to it.

therealvio committed:
chore: update structlog dependenices

It's been a while since this branch and related item of work has been
picked up. So we may as well update the dependencies and do a
soft-restart here.

therealvio committed:
style: capital letters for start of commentary

therealvio committed:
docs: clarify intention of slogger variable

therealvio committed:
fix: route table id is being updated with nat gateway id

While updating the logging to switch to structured logging, feedback was
provided that identified a case of inaccurate and broken logging when
`replace_route()` is invoked. Not only is the log message wrong by using
a route table id instead of the route id, but also the use of the
`new_route_table` variable would produce a dictionary in the log.

The log line has now been updated to accurately state what is going on,
and remove the use of a nested dictionary in the structured log.

BREAKING CHANGE: Users monitoring this message with the original
specific phrasing will experience broken monitors. It is recommended
that users should be monitoring generally for ERROR level logs,
or the "Failed connectivity tests! Replacing route" logs specifically.

bwhaley commented on .gitignore:
```diff
@@ -5,3 +5,10 @@
+devbox.lock
+.devbox
+.envrc
+/functions/replace-route/setup
```
Added by mistake?

therealvio replied:
Whoops! The offending commit has been fixed, and force-pushed.

bwhaley commented on functions/replace-route/app.py:
```diff
@@ -345,19 +372,19 @@ def get_current_nat_instance_id(asg_name):
             if instance['LifecycleState'] == 'InService':
                 return instance['InstanceId']
     except Exception as e:
-        logger.error(f"Failed to retrieve NAT instance ID from ASG {asg_name}: {e}")
+        slogger.error(f"Failed to retrieve NAT instance ID from ASG {asg_name}: {e}")
         return None
 
 def connectivity_test_handler(event, context):
     if not isinstance(event, dict):
-        logger.error(f"Unknown event: {event}")
+        slogger.error("Unknown event: %s", {event})
         return
 
     if event.get("source") != "aws.events":
-        logger.error(f"Unable to handle unknown event type: {json.dumps(event)}")
+        slogger.error("Unable to handle unknown event type: %s", json.dumps(event))
```
How does this appear in the logs? Is the JSON value of `event` legible in the output? This seems like an unlikely error, and I'm not sure how to easily test it, but I would be curious if it's readable.

bwhaley reviewed (commented):
Left a couple more comments. Kicking off tests.

bwhaley reviewed (commented):
The addition of `structlog` and `orjson` and `python-json-logger` actually adds a fair amount of bloat to the package. Currently, only `boto3` is needed, which is included by default in the Lambda runtime. The others are not, so they need to be added as a layer to the Lambda function.

How did you test it?

therealvio committed:
fix: use json-logger to for stdlib logging in deps
